cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,327 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Heuristic qubit routing algorithm based on arxiv:1902.08091."""
+
+from typing import List, Optional, Set, Tuple, TYPE_CHECKING
+from itertools import combinations
+import networkx as nx
+
+from cirq import circuits, ops, protocols
+from cirq.transformers import transformer_api, transformer_primitives
+from cirq.transformers.routing import mapping_manager, line_initial_mapper
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@transformer_api.transformer
+class RouteCQC:
+    """Transformer class that implements a circuit routing algorithm.
+
+    The algorithm proceeds as follows:
+        (1) Compute the timesteps of the circuit: considering operations in the given circuit from
+            beginning to end, the next timestep is a maximal set of 2-qubit operations that act on
+            disjoint qubits. It is 'maximal' because any 2-qubit gate's qubits in the next timestep
+            must intersect with the qubits that are acted on in the current timestep.
+
+        (2) Places the logical qubits in the input circuit onto some input device by using an
+            initial mapper (LineInitialMapper by default).
+
+        (3) Insert necessary swaps to ensure all 2-qubit gates are executable on the device by
+            traversing the timesteps from left to right while:
+                (i) Removing any single qubit gate and executable 2-qubit gate in the current
+                    timestep and add it to the output routed circuit.
+               (ii) If there aren't any gates left in the current timestep, move on to the next.
+              (iii) If there are gates remaining in the current timesteps, consider a set of
+                    candidate swaps on them and rank them based on a heuristic cost function. Pick
+                    the swap that minimises the cost and use it to update our logical to physical
+                    mapping. Repeat from (i).
+    """
+
+    def __init__(self, device_graph: nx.Graph):
+        """Initializes the circuit routing transformer."""
+
+        if nx.is_directed(device_graph):
+            raise ValueError("Device graph must be undirected.")
+        self.device_graph = device_graph
+
+    def __call__(
```

Comment:
do we want to use `__call__` for this? Why not a method name?

If you do want call: why can't this class just be a function?

Reply:
`__call__` needs to exist because it's a cirq transfomer and `cirq.TRANSFORMER` API requires every transformer to be "a callable that can take an abstract circuit + transformer context and return an abstract circuit" along with additional keyword arguments.

Reply:
got it. thanks!

Reply:
We might want to keep this transformer as a class because we want to make the cost function `_cost(self)` overwrite-able by other implementations of this routing algorithm. A example use case for this would be implementing noise-aware cost function that penalizes swaps on edges that have higher cost than others.